### ISSUE TITLE
Factories with relationship fields are calling create() when calling make()

### ISSUE THREAD

**Issue by robertoarruda** [laravel/framework#41313]
### Discussed in https://github.com/laravel/framework/discussions/36169

<div type='discussions-op-text'>

<sup>Originally posted by **robertoarruda** February  6, 2021</sup>
- Laravel Version: 8.22.0
- PHP Version: 7.4.13

### Description:
I use factories in my unit tests. My tests never touch the database, so I don't use connections or migrations on it.
In my unit tests, factories use the make method only to provide fake data. However, even using the make method, apparently the related factories are using the create method.
My code has several relationships in the factories, but below is a simple example of the problem.

### Steps To Reproduce:
PositionFactory:
```php
<?php

namespace Database\Factories;

use Illuminate\Database\Eloquent\Factories\Factory;
use App\Models\Firm;
use App\Models\Position;

class PositionFactory extends Factory
{
    /**
     * @var string
     */
    protected $model = Position::class;

    /**
     * @return array
     */
    public function definition(): array
    {
        return [
            'name' => $this->faker->name,
            'description' => $this->faker->text,
            'firm_id' => Firm::factory(),
        ];
    }
}

``` 
FirmFactory:
```php
<?php

namespace Database\Factories;

use Illuminate\Database\Eloquent\Factories\Factory;
use App\Models\Firm;

class FirmFactory extends Factory
{
    /**
     * @var string
     */
    protected $model = Firm::class;

    /**
     * @return array
     */
    public function definition(): array
    {
        return [
            'name' => $this->faker->company,
            'description' => $this->faker->paragraph,
            'website' => $this->faker->url,
            'discount' => $this->faker->randomFloat(),
            'supplier' => $this->faker->word,
            'status' => $this->faker->randomElement(['active', 'disabled']),
        ];
    }
}

``` 

The error presented is when running the tests, it is the attempt to create a user. I imagine it’s because most relationships are tied to that table
```shell
  • Tests\UT\Repositories\TemplateFieldPropertyValueRepositoryTest > attach dependency existing dependency
   Illuminate\Database\QueryException 

  SQLSTATE[HY000]: General error: 1 no such table: users (SQL: insert into "users" ("first_name", "last_name", "email", "password", "secret_phrase", "updated_at", "created_at") values (Fritz, Murray, tania32@stroman.org, $2y$10$Tc6AvPJ3Ijsh2kObc708NuCjscMlgOkrH8yxaBD9CBJZLSiF48pIC, Quis occaecati optio suscipit quia rerum esse. Sapiente voluptatem molestias temporibus repudiandae et. Illo at distinctio cum corporis placeat. Sed quam occaecati sed non., 2021-02-06 23:24:37, 2021-02-06 23:24:37))

  at vendor/illuminate/database/Connection.php:678
    674▕         // If an exception occurs when attempting to run a query, we'll format the error
    675▕         // message to include the bindings with SQL, which will make this exception a
    676▕         // lot more helpful to the developer instead of just the database's errors.
    677▕         catch (Exception $e) {
  ➜ 678▕             throw new QueryException(
    679▕                 $query, $this->prepareBindings($bindings), $e
    680▕             );
    681▕         }
    682▕ 

  1   [internal]:0
      Illuminate\Database\Eloquent\Factories\Factory::Illuminate\Database\Eloquent\Factories\{closure}(Object(Database\Factories\UserFactory), "user_id")

  2   [internal]:0
      Illuminate\Database\Eloquent\Factories\Factory::Illuminate\Database\Eloquent\Factories\{closure}(Object(Database\Factories\TemplateFactory), "template_id")


  Tests:  2 failed
  Time:   0.17s

```
Debugging the code to solve the problem, I found the place where the create method is called.
[Illuminate/Database/Eloquent/Factories/Factory.php](https://github.com/laravel/framework/blob/v8.22.0/src/Illuminate/Database/Eloquent/Factories/Factory.php#L413)
```php
    /**
     * Expand all attributes to their underlying values.
     *
     * @param  array  $definition
     * @return array
     */
    protected function expandAttributes(array $definition)
    {
        return collect($definition)->map(function ($attribute, $key) use (&$definition) {
            if (is_callable($attribute) && ! is_string($attribute) && ! is_array($attribute)) {
                $attribute = $attribute($definition);
            }

            if ($attribute instanceof self) {
                $attribute = $attribute->create()->getKey(); // <======= HERE
            } elseif ($attribute instanceof Model) {
                $attribute = $attribute->getKey();
            }

            $definition[$key] = $attribute;

            return $attribute;
        })->all();
    }
```
I believe this code should check first, whether it is called by the make or create method, and then use the correct method.
(If I switch to the make method, the tests work).</div>

**Comment by driesvints** [comment]
Heya, thanks for reporting.

We'll need more info and/or code to debug this further. Can you please create a repository with the command below, commit the code that reproduces the issue as separate commits on the main/master branch and share the repository here? Please make sure that you have the [latest version of the Laravel installer](https://github.com/laravel/installer) in order to run this command. Please also make sure you have both Git & [the GitHub CLI tool](https://cli.github.com) properly set up.

```bash
laravel new bug-report --github="--public"
```

Please do not amend and create a separate commit with your custom changes. After you've posted the repository, we'll try to reproduce the issue.

Thanks!

**Comment by driesvints** [comment]
I'll re-open this once you managed to provide a repo. Thanks

**Comment by erichowey** [comment]
I am also having this issue. I created an example app to reproduce this issue:

https://github.com/erichowey/LaravelExample

```
eric@beast:~/example-app$ ./artisan tinker
Psy Shell v0.11.2 (PHP 8.0.17 — cli) by Justin Hileman
>>> App\Models\Post::factory()->make()
=> App\Models\Post {#3557
     user_id: 1,
     post: "Sed et rerum exercitationem in. Est mollitia repellat et consectetur sit. Repellendus modi et recusandae a ad.",
   }
>>> 
```

As you can see, a user was persisted even though `App\Models\Post::factory()->make()` was called
![image](https://user-images.githubusercontent.com/4033313/160174985-2284583b-88e8-4e93-adad-c74563013148.png)

However, the post was not persisted as expected.
![image](https://user-images.githubusercontent.com/4033313/160175262-25abdb34-3916-493a-8540-30a73444de69.png)

**Comment by driesvints** [comment]
@erichowey please recreate that repo with the command I posted.

**Comment by erichowey** [comment]
@driesvints Here you go:
https://github.com/erichowey/laravel-bug-report

Here's a PR showcasing my commits to the repo:
https://github.com/erichowey/laravel-bug-report/pull/1/files

**Comment by driesvints** [comment]
So I don't think this is a bug. You define a `user_id` in your factory. This ID cannot be created without actually creating the model.

**Comment by robertoarruda** [comment]
This is recommended in the documentation.
https://laravel.com/docs/9.x/database-testing#defining-relationships-within-factories

Also, don't you also think it doesn't feel right to use the factory one way to `create()` and another way to `make()`?

**Comment by robertoarruda** [comment]
> So I don't think this is a bug. You define a `user_id` in your factory. This ID cannot be created without actually creating the model.

This is recommended in the documentation.
https://laravel.com/docs/9.x/database-testing#defining-relationships-within-factories

Also, don't you also think it doesn't feel right to use the factory one way to `create()` and another way to `make()`?

**Comment by driesvints** [comment]
> This is recommended in the documentation.

Yes that's how you define relationships. I don't see your point?

> Also, don't you also think it doesn't feel right to use the factory one way to create() and another way to make()?

I don't understand what you mean.

**Comment by erichowey** [comment]
> So I don't think this is a bug. You define a `user_id` in your factory. This ID cannot be created without actually creating the model.

Wouldn't it make more sense to not persist the user but immediately delete it afterwards to prevent stale records?

**Comment by erichowey** [comment]
> @erichowey I don't see why you'd want that. You're defining a `user_id` as a required field in your factory. It needs to exist for the state of the model to be complete.

@driesvints Because I'm using `make()` not `create()`. Per the documentation:
> we'll use the make method to create models without persisting them to the database:

As of right now `make()` creates the model without persisting that model to the database, however, any relational models within the factory are persisted. In my opinion, this doesn't make sense when using make and leads to stale records in the database. `make()` should do absolutely not persistence to the database. This includes relationships.

**Comment by robertoarruda** [comment]
> > This is recommended in the documentation.
> 
> Yes that's how you define relationships. I don't see your point?
> 
> > Also, don't you also think it doesn't feel right to use the factory one way to create() and another way to make()?
> 
> I don't understand what you mean.

Unit tests shouldn't touch the database, so it's necessary that the factory doesn't create anything in the database (in my unit tests the database doesn't even exist).
Best practices in unit testing advise that a test should only test a single logical unit and simulate everything that comes out of its unit.

This is the opposite of integration testing, which needs to fully test a feature.

For this reason, we need to define relationships in the factory so that they can be used in integration tests (`Model::factory()->create()`), but we need this to not try to create database data in unit tests (`Model::factory()->make()`).

**Comment by driesvints** [comment]
So, what I'm trying to say is that there's no way to create the identifier without the database. It's an incremental ID. It gets generated by the database. There's no other mechanism that generates identifiers. What do you expect to happen for this value if no database is involved?

**Comment by robertoarruda** [comment]
> So, what I'm trying to say is that there's no way to create the identifier without the database. It's an incremental ID. It gets generated by the database. There's no other mechanism that generates identifiers. What do you expect to happen for this value if no database is involved?

In my opinion, when `make()` is called on a factory with relationships defined on the foreign key, it should ignore the field and return a null on the foreign key.
The developer is the one who should handle this issue on a case-by-case basis.

Example:
```php
Post::factory()->make();

// returns
// [
//     'user_id' => null,
//     'post' => 'random text',
// ];
```
```php
Post::factory()->make(['user_id' => 123456]);

// returns
// [
//     'user_id' => 123456,
//     'post' => 'random text',
// ];
```

**Comment by erichowey** [comment]
> What do you expect to happen for this value if no database is involved?

That it doesn't work. However, when a database is involved, I expect it to create the record to get the auto increment and then delete it when done because we're using `make()` not `create()`.

**Comment by driesvints** [comment]
> However, when a database is involved, I expect it to create the record to get the auto increment and then delete it when done because we're using make() not create().

I don't think you can expect that to actually happen, sorry. To me, that would even make less sense than not making the model at all.

**Comment by erichowey** [comment]
@driesvints That's fine, however, the fact remains that `make()` is causing data to be persisted to the database when the documentation says that data should only be persisted when using `create()`. Relation or not, the fact that any data is persisted to the database when using `make()` doesn't make sense to me.

**Comment by robertoarruda** [comment]
> > So, what I'm trying to say is that there's no way to create the identifier without the database. It's an incremental ID. It gets generated by the database. There's no other mechanism that generates identifiers. What do you expect to happen for this value if no database is involved?
> 
> In my opinion, when `make()` is called on a factory with relationships defined on the foreign key, it should ignore the field and return a null on the foreign key. The developer is the one who should handle this issue on a case-by-case basis.
> 
> Example:
> 
> ```
> Post::factory()->make();
> 
> // returns
> // [
> //     'user_id' => null, // or 0
> //     'post' => 'random text',
> // ];
> ```
> 
> ```
> Post::factory()->make(['user_id' => 123456]);
> 
> // returns
> // [
> //     'user_id' => 123456,
> //     'post' => 'random text',
> // ];
> ```

@driesvints does this make sense to you?

**Comment by robertoarruda** [comment]
> @robertoarruda no I think the current behavior makes sense.
> 
> I'm sorry guys but I'm going to leave it at this. Welcoming PR's to clarify things in the docs.

So, in your opinion, is it correct to create a new user in the database, when the code below is run in production?
```php
$post = Post::factory()->make();
```

**Comment by driesvints** [comment]
@robertoarruda yes. And please stop downvoting me. If you want to change anything then step up and attempt a PR.

**Comment by robertoarruda** [comment]
> @robertoarruda yes. And please stop downvoting me. If you want to change anything then step up and attempt a PR.

Sorry @driesvints, I don't mean to sound boring and rude to you. I'm just insisting on this, because the issue was closed for lack of more information. The information has been provided, but you are rejecting the issue we are complaining about.

We need this behavior to be accepted as an issue so I can do a PR to solve it, or even give a better developer than me the opportunity to solve this problem.

**Comment by driesvints** [comment]
> We need this behavior to be accepted as an issue so I can do a PR to solve it, or even give a better developer than me the opportunity to solve this problem.

@robertoarruda anyone can send in a PR. A closed issue isn't stopping you from doing that.

>  I apologize, but this was a disappointing experience.

Look, I'm sorry you feel that way but in the end it's just us disagreeing that this is an issue in the first place. If you can't accept that there's going to be differences in opinion in open source or anything in life really then you're going to be disappointed a lot more.

> I've personally given my idea of how to resolve this.

I've also tried to do my best to explain why this isn't an issue and why this works the way that it does. I'm sorry but I'm just doing my best to maintain this project and I feel that this works as intended. If you don't agree then please step up and send in a PR to prove that it can be solved differently because I don't think it can.

> Any PR I would attempt would be in vain because myself and my team simply don't have any better ideas.

Doesn't this exactly proves that it can't be solved?

> The correct solution would be to leave this bug open and allow the community to view it and hopefully someone smarter than me can come up with a good idea to this problem.

That's not how we handle issues we don't consider bugs, sorry.

> you stonewalled us

How? I only told you my opinion about it as a maintainer. I said we're open to PRs. You keep insisting this is a bug while it isn't.

> I've heard rumors of Laravel's unhealthy obsession with 0 bugs, but now I've experienced it first hand. I can't help but feel a little frustrated.

I'm truly sorry you feel that way. But in my defence I can say the same about you keeping on insisting that there's a problem. And if you call our goal about solving real issues to make sure there's as few bugs and open tickets an "obsession" then I just think you don't get our work ethic 😅 

> On Laravel-side, it's not a easy-peasy patch tho, because make() is called also from create(), so we should carry a boolean to tell expandAttributes() if to create or not the related factory.

@danilopolani yeah that could indeed get a bit messy. Appreciate the deep dive on this though!

> If you just need to work on the relation ID, just set it manually with make(['author_id' => 123]).

This is a good solution for some cases and now that I see it I think I also used this in the past.

**Comment by erichowey** [comment]
@danilopolani Thank you for the explanation and taking the time to investigate this. For those of us who aren't seasoned developers, it's extremely helpful. I'll attempt a PR to the docs with your explanation tomorrow.

**Comment by robertoarruda** [comment]
I tried a PR that solves this problem, but it was refused.
Unfortunately, we will have to live with this incorrect behavior.

**Comment by erichowey** [comment]
My doc PR was closed as well with no feedback ¯\_(ツ)_/¯

**Comment by tomb1n0** [comment]
Just weighing in on this.

I've moved over to using UUID's for my models primary keys recently, which means i absolutely do _NOT_ need to do an insert to determine the primary key of my nested relationships.

**Comment by rreynier** [comment]
Not a big fan of how this behaves either. If the purpose of  `make()` is to NOT use the database, it seems like somehow, someway it should not try to insert data into the database. This was completely unexpected and incorrect behavior IMO.  I do recognize the challenges posted above, but to me the question is which solution is less wrong. Way I see it just returning null without hitting the database is the lesser of two evils when using `make()`.

**Comment by madebycaliper** [comment]
Hey all 👋 I think this is a great opportunity to improve the DX for testing with Laravel. I appreciate the dialog here and hope there's continued exploration of this issue.

I'm not a super-seasoned Laravel dev and don't have a proposed solution, but [this suggestion](https://github.com/laravel/framework/discussions/36169#discussioncomment-4207175) feels like "the Laravel way" and could at least offer a convenience method to force `make()` behave the way it's expected it to.

I understand that the Laravel core team might not see eye-to-eye on how `make()` is "supposed to work" but that seems to be the way it's explained in the docs and how I've always understood it in the 6 years I've been using Laravel. Maybe the docs just require some more illustrations of how to use `Factory::make()` when there are related models that get generated? 🤷‍♂️

**Comment by Spotnyk** [comment]
@driesvints  Quick clarifying question (for myself). Reading through this thread (as my team is dealing with this), I've noticed you say that this is not a problem, _and_ that there is no solution.

Those are not the same thing.

Does the core team not consider this an issue to begin with, or does the core team not know how to deal with said issue? Again; curious.

Appreciate your time.

**Comment by kirkbushell** [comment]
To be honest, not creating related records with autoincrement ids as part of tests is fraught with problems - you will undoubtedly encounter issues, and the hair-pulling that would result is worse than the data being created in the database to begin with. 

However, there is certainly a missed case here - UUIDs and ULIDs. These IDs are normally created before the model is persisted, meaning that you have the IDs available for testing - make() should respect this when it's present - and I believe that is the real issue.

For autoincrement ids, if you want to make related objects, your best bet is to set the relationship on the model manually, by making the relevant model objects and then using setRelation (for now). That's what we're doing with ours, but it would be great if the factory system was able to recognise IDs that are not auto-incremented and can therefore rely on make() over create() to initialise those related objects.

**Comment by logic-b-o-m-b** [comment]
Laravel factories are not suitable for use in unit tests for two main reasons: they are tightly coupled with the database, and they are extremely slow compared to other options.

I wrote a HasFixture trait to add a fixture() method to factory classes. I used it instead of make() to achieve what make should actually do. It worked great; primary keys and foreign keys had correct values (ints/UUIDs). However, it was too slow, so I gave up on it.

What I usually do is:

Easy way (or at least, what should be easy):
I strip everything that involves I/O from the classes or set of classes I want to unit test. I then test just the business/application logic without making any I/O calls.

Hard way:

I implement Ports & Adapters.

I have a Repository where I keep queries (either a repository per model, or per model and operation type — read/write as in CQRS, meaning two repositories for one model).

Repositories are driven adapters implementing ports.

I have a service provider with a binding for each repository.

Repositories return non-scalar values, simple arrays, or DTOs for models with nested models (e.g., using spatie/laravel-data).

I have fake implementations of repositories, which implement a contract and track changes made to data for easy assertions.

To easily generate test fixtures, fake repositories sometimes use fixture builders or static fixtures.

I'm an architecture freak, so I don't mind boilerplate overhead, and I accept that I have to update builders of static fixtures from time to time, especially that my tests are super fast.
